lib/lambda_ethereum_consensus/p2p/gossip/operations_collector.ex

```diff
@@ -202,68 +219,94 @@ defmodule LambdaEthereumConsensus.P2P.Gossip.OperationsCollector do
   end
 
   # TODO: filter duplicates
-  defp handle_msg({operation, msg}, state)
+  defp handle_msg({operation, msg})
        when operation in @operations do
-    new_msgs = [msg | Map.fetch!(state, operation)]
-    {:noreply, Map.replace!(state, operation, new_msgs)}
+    new_msgs = [msg | fetch_operation!(operation)]
+    store_operation(operation, new_msgs)
+    {:noreply, nil}
   end
 
-  defp filter_messages(state, slot, operations) do
+  defp filter_messages(slot, operations) do
     indices =
       operations.bls_to_execution_changes
       |> MapSet.new(& &1.message.validator_index)
 
-    bls_to_execution_changes =
-      state.bls_to_execution_change
-      |> Enum.reject(&MapSet.member?(indices, &1.message.validator_index))
+    fetch_operation!(:bls_to_execution_change)
+    |> Enum.reject(&MapSet.member?(indices, &1.message.validator_index))
+    |> then(&store_operation(:bls_to_execution_change, &1))
 
     # TODO: improve AttesterSlashing filtering
-    attester_slashings =
-      state.attester_slashing |> Enum.reject(&Enum.member?(operations.attester_slashings, &1))
+    fetch_operation!(:attester_slashing)
+    |> Enum.reject(&Enum.member?(operations.attester_slashings, &1))
+    |> then(&store_operation(:attester_slashing, &1))
 
     slashed_proposers =
       operations.proposer_slashings |> MapSet.new(& &1.signed_header_1.message.proposer_index)
 
-    proposer_slashings =
-      state.proposer_slashing
-      |> Enum.reject(
-        &MapSet.member?(slashed_proposers, &1.signed_header_1.message.proposer_index)
-      )
+    fetch_operation!(:proposer_slashing)
+    |> Enum.reject(&MapSet.member?(slashed_proposers, &1.signed_header_1.message.proposer_index))
+    |> then(&store_operation(:proposer_slashing, &1))
 
     exited = operations.voluntary_exits |> MapSet.new(& &1.message.validator_index)
 
-    voluntary_exits =
-      state.voluntary_exit |> Enum.reject(&MapSet.member?(exited, &1.message.validator_index))
+    fetch_operation!(:voluntary_exit)
+    |> Enum.reject(&MapSet.member?(exited, &1.message.validator_index))
+    |> then(&store_operation(:voluntary_exit, &1))
 
     # TODO: improve attestation filtering
     added_attestations = MapSet.new(operations.attestations)
 
-    attestations =
-      state.attestation
-      |> Stream.reject(&MapSet.member?(added_attestations, &1))
-      |> Enum.reject(&old_attestation?(&1, slot))
-
-    %{
-      state
-      | bls_to_execution_change: bls_to_execution_changes,
-        attester_slashing: attester_slashings,
-        proposer_slashing: proposer_slashings,
-        voluntary_exit: voluntary_exits,
-        attestation: attestations,
-        slot: slot
-    }
+    fetch_operation!(:attestation)
+    |> Stream.reject(&MapSet.member?(added_attestations, &1))
+    |> Enum.reject(&old_attestation?(&1, slot))
+    |> then(&store_operation(:attestation, &1))
```

Comment:
Aren't we missing an update for the slot here? Or is it not necessary? If it's not necessary we should not save it at all. If it is, we should add a test that reflects its importance.

Reply:
Actually yes, nice catch! I will correct it and add some tests for that.